File: jEverest/src/ca/marc/everest/datatypes/generic/SLIST.java
/* 
 * Copyright 2008/2011 Mohawk College of Applied Arts and Technology
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you 
 * may not use this file except in compliance with the License. You may 
 * obtain a copy of the License at 
 * 
 * http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 
 * License for the specific language governing permissions and limitations under 
 * the License.
 * 
 * User: Justin Fyfe
 * Date: 10-01-2012
 */
package ca.marc.everest.datatypes.generic;

import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;

import ca.marc.everest.annotations.*;
import ca.marc.everest.datatypes.*;
import ca.marc.everest.datatypes.interfaces.*;

/**
 * Represents a list of sampled values with each new term 
 * scaled and translated from a list of previous samples. Used
 * to specify sampled biosignals
 */
@Structure(name = "SLIST", structureType = StructureType.DATATYPE)
public class SLIST<T extends IQuantity> extends ANY implements ISequence<INT>, ISampledList {

	// backing field for digits
	private LIST<INT> m_digits = new LIST<INT>();
	// backing field for the origin property
	private T m_origin;
	// backing field for the scale property
	private IQuantity m_scale;
	
	/**
	 * Creates a new instance of the sampled list class
	 */
	public SLIST() { super(); }
	/**
	 * Creates a new instance of the sampled list class with the specified
	 * origin and scale
	 * @param origin The origin (first reading in the list) a physical quantity which a zero-digit in the sequence would represent
	 * @param scale The ratio-scale quantity that is factored out of the digit sequence
	 */
	public SLIST(T origin, IQuantity scale) { 
		super();
		this.m_origin = origin;
		this.m_scale = scale;
	}
	/**
	 * Creates a new instance of the sampled list class with the specified origin, scale
	 * and sequence items.
	 * @param origin The origin (first reading in the list) which represents a zero-digit in the sequence
	 * @param scale The ratio-scale quantity that is factored out of the digit sequence
	 * @param items The digits representing the readings or samples
	 */
	public SLIST(T origin, IQuantity scale, Collection<INT> items)
	{
		this(origin, scale);
		this.m_digits = new LIST<INT>(items);
	}
	/**
	 * Creates a new sampled list class
	 * @param origin The origin (first reading in the list) which represents a zero-digit in the sequence
	 * @param scale The ratio-scale quantity that is factored out of the digit sequence
	 * @param items The digits representing the readings or samples
	 * @return A new instance of SLIST
	 */
	public static <T extends IQuantity> SLIST<T> CreateSLIST(T origin, IQuantity scale, INT... items)
	{
		return new SLIST<T>(origin, scale, Arrays.asList(items));
	}

	/**
	 * Gets the origin of the sampled list
	 */
	@Property(name = "origin", propertyType = PropertyType.STRUCTURAL, conformance = ConformanceType.REQUIRED)
	@Override
	public IQuantity getOrigin() {
		return this.m_origin;
	}

	/**
	 * Sets the origin of the sampled list
	 */
	@SuppressWarnings("unchecked")
	@Override
	public void setOrigin(IQuantity value) {
		this.m_origin = (T)value;
	}

	/**
	 * Gets the ratio-scale quantity that is factored out of the digit
	 */
	@Property(name = "scale", propertyType = PropertyType.STRUCTURAL, conformance = ConformanceType.REQUIRED)
	@Override
	public IQuantity getScale() {
		return this.m_scale;
	}

	/**
	 * Sets the ratio-scale quantity that is factored out of the digit
	 */
	@Override
	public void setScale(IQuantity value) {
		this.m_scale = value;
	}

	/**
	 * Gets the items in the list
	 */
	@Property(name = "digit", conformance = ConformanceType.REQUIRED, propertyType = PropertyType.STRUCTURAL)
	@Override
	public Collection<INT> getItems() {
		return this.m_digits.getItems();
	}

	/**
	 * Gets the first sample in the list
	 */
	@Override
	public INT first() {
		return this.m_digits.first();
	}

	/**
	 * Gets the last sample from the list
	 */
	@Override
	public INT last() {
		return this.m_digits.last();
	}

	/**
	 * Selects a sub-sequence from the sampled list
	 */
	@Override
	public ISequence<INT> subSequence(int start, int end) {
		return this.m_digits.subSequence(start, end);
	}

	/**
	 * Selects a sub=sequence from the sampled list
	 */
	@Override
	public ISequence<INT> subSequence(int start) {
		return this.subSequence(start);
	}
	
	/**
	 * Get the specified item at the index
	 */
	@Override
	public INT get(int index) {
		return this.m_digits.get(index);
	}
	/**
	 * Returns true if this sampled list contains all the samples in other
	 */
	@Override
	public BL includesAll(ICollection<INT> other) {
		return this.m_digits.includesAll(other);
	}
	/**
	 * Returns true if this sampled list contains none of the samples in other
	 */
	@Override
	public BL excludesAll(ICollection<INT> other) {
		return this.m_digits.excludesAll(other);
	}
	/**
	 * Returns true if the sampled list contains no samples
	 */
	@Override
	public boolean isEmpty() {
		return this.m_digits.isEmpty();
	}
	/**
	 * Gets the iterator for this item
	 */
	@Override
	public Iterator<INT> iterator() {
		return this.m_digits.iterator();
	}
	/**
	 * Add a sample to this SLIST
	 */
	@Override
	public boolean add(INT e) {
		return this.m_digits.add(e);
	}
	/**
	 * Adds all items in the collection to this sampled list
	 */
	@Override
	public boolean addAll(Collection<? extends INT> c) {
		return this.m_digits.addAll(c);
	}
	/**
	 * Clears all items from this list
	 */
	@Override
	public void clear() {
		this.m_digits.clear();
	}
	/**
	 * Returns true if the sampled list contains the object
	 */
	@Override
	public boolean contains(Object o) {
		return this.m_digits.contains(o);
	}
	/**
	 * Returns true if the sampled list contains all the items in the specified collection
	 */
	@Override
	public boolean containsAll(Collection<?> c) {
		return this.m_digits.containsAll(c);
	}
	/**
	 * Removes the specified sample from the list of samples
	 */
	@Override
	public boolean remove(Object o) {
		return this.m_digits.remove(o);
	}
	/**
	 * Removes all the samples contained in c from the list of samples
	 */
	@Override
	public boolean removeAll(Collection<?> c) {
		return this.m_digits.removeAll(c);
	}
	/**
	 * Performs the retainAll function against the collection of samples
	 */
	@Override
	public boolean retainAll(Collection<?> c) {
		return this.m_digits.retainAll(c);
	}
	/**
	 * Gets the number of samples in the sample list
	 */
	@Override
	public int size() {
		return this.m_digits.size();
	}
	/**
	 * Converts this list of samples to an array
	 */
	@Override
	public Object[] toArray() {
		return this.m_digits.toArray();
	}
	/**
	 * Converts the list of samples to an array
	 */
	@SuppressWarnings({ "unchecked", "hiding" })
	@Override
	public <T> T[] toArray(T[] a) {
		return (T[])this.m_digits.toArray();
	}

	/** 
	 * Calculates the hash code of this sampled list
	 */
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = super.hashCode();
		result = prime * result
				+ ((m_digits == null) ? 0 : m_digits.hashCode());
		result = prime * result
				+ ((m_origin == null) ? 0 : m_origin.hashCode());
		result = prime * result + ((m_scale == null) ? 0 : m_scale.hashCode());
		return result;
	}
	/**
	 * Determines the value equality of this sampled list
	 */
	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (!super.equals(obj))
			return false;
		if (getClass() != obj.getClass())
			return false;
		SLIST other = (SLIST) obj;
		if (m_digits == null) {
			if (other.m_digits != null)
				return false;
		} else if (!m_digits.equals(other.m_digits))
			return false;
		if (m_origin == null) {
			if (other.m_origin != null)
				return false;
		} else if (!m_origin.equals(other.m_origin))
			return false;
		if (m_scale == null) {
			if (other.m_scale != null)
				return false;
		} else if (!m_scale.equals(other.m_scale))
			return false;
		return true;
	}
}

File: MARC.Everest/DataTypes/SLIST.cs
/* 
 * Copyright 2008-2013 Mohawk College of Applied Arts and Technology
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you 
 * may not use this file except in compliance with the License. You may 
 * obtain a copy of the License at 
 * 
 * http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 
 * License for the specific language governing permissions and limitations under 
 * the License.
 * 
 * User: Justin Fyfe
 * Date: 01-09-2009
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using MARC.Everest.DataTypes.Interfaces;
using MARC.Everest.Attributes;
using System.Collections;

namespace MARC.Everest.DataTypes
{
    /// <summary>
    /// A sequence of sampled values scaled and translated from a 
    /// list of integer values. Used to specify sampled biosignals
    /// </summary>
    /// <remarks>
    /// <para>
    /// The type of <see cref="P:Scale"/> is dictated by <typeparamref name="T"/>. For example, when
    /// a SLIST&lt;PQ> is specified, the <see cref="P:Scale"/> should be an instance of PQ.
    /// </para>
    /// <para>
    /// SLIST was introduced in DataTypes R2 and should not be used when formatting instances using 
    /// DataTypes R1 formatters.
    /// </para>
    /// </remarks>
    [Structure(Name = "SLIST", StructureType = StructureAttribute.StructureAttributeType.DataType, DefaultTemplateType = typeof(PQ))]
    public class SLIST<T> : ANY, ISequence<INT>, IEquatable<SLIST<T>>, ISampledList
        where T : IQuantity, new()
    {

        /// <summary>
        /// Digits
        /// </summary>
        private LIST<INT> m_digits = new LIST<INT>();

        /// <summary>
        /// Creates a new instance of SLIST
        /// </summary>
        public SLIST()
        {
        }
        /// <summary>
        /// Creates a new instance of SLIST with the specified <paramref name="origin"/> and
        /// <paramref name="scale"/>
        /// </summary>
        public SLIST(T origin, IQuantity scale)
        {
            this.Origin = origin;
            this.Scale = scale;
        }
        /// <summary>
        /// Creates a new instance of SLIST with the <paramref name="origin"/>, <paramref name="scale"/>
        /// and <paramref name="digits"/>
        /// </summary>
        public SLIST(T origin, IQuantity scale, IEnumerable<INT> items) : this(origin, scale)
        {
            this.m_digits = new LIST<INT>(items);
        }

        /// <summary>
        /// Creates an instance of the SLIST with the specified <paramref name="items"/>
        /// </summary>
        /// <param name="items"></param>
        /// <returns></returns>
        public static SLIST<T> CreateSLIST(T origin, IQuantity scale, params INT[] items)
        {
            return new SLIST<T>(origin, scale, items);
        }

        /// <summary>
        /// The origin of the first reading in the list. The physical
        /// quantity that a zero-digit in the sequence would represent
        /// </summary>
        [Property(Name = "origin", Conformance = PropertyAttribute.AttributeConformanceType.Required, PropertyType = PropertyAttribute.AttributeAttributeType.Structural)]
        public T Origin { get; set; }

        /// <summary>
        /// Gets or sets the a ratio-scale quantity that is factored out of the digit
        /// sequence
        /// </summary>
        [Property(Name = "scale", Conformance = PropertyAttribute.AttributeConformanceType.Required, PropertyType = PropertyAttribute.AttributeAttributeType.Structural)]
        public IQuantity Scale { get; set; }

        #region IEnumerable<INT> Members

        /// <summary>
        /// Get the enumerator from the digits
        /// </summary>
        public IEnumerator<INT> GetEnumerator()
        {
            if (this.m_digits != null)
                return this.m_digits.GetEnumerator();
            return null;
        }

        #endregion

        #region IEnumerable Members

        /// <summary>
        /// Get the enumerator from the digits
        /// </summary>
        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        #endregion

        #region ISequence<INT> Members

        /// <summary>
        /// Gets the first item in the list
        /// </summary>
        public INT First
        {
            get { return this.m_digits.First; }
        }

        /// <summary>
        /// Gets the last item in the list
        /// </summary>
        public INT Last
        {
            get { return this.m_digits.Last; }
        }

        /// <summary>
        /// Returns a sub-sequence from the <see cref="P:Digits"/>
        /// property
        /// </summary>
        /// <param name="start">The first item to retrieve</param>
        /// <param name="end">The last item in the list to retrieve</param>
        /// <returns>A new instance of <see cref="T:LIST"/> containing the results</returns>
        public ISequence<INT> SubSequence(int start, int end)
        {
            return this.m_digits.SubSequence(start, end);
        }

        /// <summary>
        /// Returns a sub-sequence from the <see cref="P:Digits"/>
        /// property starting at <paramref name="start"/> and 
        /// ending with the length of the list
        /// </summary>
        /// <param name="start"></param>
        /// <returns>A new instance of <see cref="T:LIST"/> containing the results</returns>
        public ISequence<INT> SubSequence(int start)
        {
            return this.m_digits.SubSequence(start);
        }

        #endregion

        #region IColl<INT> Members

        /// <summary>
        /// Gets a list of all items in the SLIST
        /// </summary>
        [Property(Name = "digit", Conformance = PropertyAttribute.AttributeConformanceType.Required, PropertyType = PropertyAttribute.AttributeAttributeType.NonStructural)]
        public List<INT> Items
        {
            get { return this.m_digits.Items; }
        }

        /// <summary>
        /// Returns a new instance of <see cref="T:BL"/> indicating if
        /// this SLIST contains all the items from <paramref name="other"/>
        /// </summary>
        /// <param name="other">The other collection to detect inclusion</param>
        public BL IncludesAll(IColl<INT> other)
        {
            return this.m_digits.IncludesAll(other);
        }

        /// <summary>
        /// Returns  a new instance of <see cref="T:BL"/> indicating whether
        /// the SLIST excludes all the items from <paramref name="other"/>
        /// </summary>
        /// <param name="other">The other collection to detect exclusion</param>
        public BL ExcludesAll(IColl<INT> other)
        {
            return this.m_digits.ExcludesAll(other);
        }

        #endregion

        #region IList<INT> Members

        /// <summary>
        /// Return the index of <paramref name="item"/> within this SLIST
        /// </summary>
        /// <param name="item">The item for which a search should be conducted</param>
        public int IndexOf(INT item)
        {
            return this.m_digits.IndexOf(item);
        }

        /// <summary>
        /// Inserts <paramref name="item"/> at the specified <paramref name="index"/>
        /// </summary>
        public void Insert(int index, INT item)
        {
            this.m_digits.Insert(index, item);
        }

        /// <summary>
        /// Remove the object at <paramref name="index"/>
        /// </summary>
        public void RemoveAt(int index)
        {
            this.m_digits.RemoveAt(index);
        }

        /// <summary>
        /// Indexer property for the SLIST
        /// </summary>
        public INT this[int index]
        {
            get
            {
                return this.m_digits[index];
            }
            set
            {
                this.m_digits[index] = value;
            }
        }

        #endregion

        #region ICollection<INT> Members

        /// <summary>
        /// Adds <paramref name="item"/> to the SLIST
        /// </summary>
        public void Add(INT item)
        {
            this.m_digits.Add(item);
        }

        /// <summary>
        /// Clears the current instance of SLIST of all readings
        /// </summary>
        public void Clear()
        {
            this.m_digits.Clear();
        }

        /// <summary>
        /// Indicates whether this instance of SLIST contains <paramref name="item"/>
        /// </summary>
        public bool Contains(INT item)
        {
            return this.m_digits.Contains(item);
        }

        /// <summary>
        /// Copy the contents of this SLIST into an array of <see cref="T:INT"/>
        /// </summary>
        /// <param name="array">The target array of the copy operation</param>
        /// <param name="arrayIndex">The index to start the copy operation</param>
        public void CopyTo(INT[] array, int arrayIndex)
        {
            this.m_digits.CopyTo(array, arrayIndex);
        }

        /// <summary>
        /// Gets the number of readings in the SLIST
        /// </summary>
        public int Count
        {
            get { return this.m_digits.Count; }
        }

        /// <summary>
        /// Indicates whether this SLIST is read only
        /// </summary>
        public bool IsReadOnly
        {
            get { return false; }
        }

        /// <summary>
        /// Removes <paramref name="item"/> from the SLIST instance
        /// </summary>
        public bool Remove(INT item)
        {
            return this.m_digits.Remove(item);
        }

        #endregion

        #region IColl Members

        /// <summary>
        /// Returns true if the SLIST is empty
        /// </summary>
        public bool IsEmpty
        {
            get { return this.m_digits.IsEmpty; }
        }

        #endregion

        #region IEquatable<SLIST<T>> Members

        /// <summary>
        /// Determine equality
        /// </summary>
        public bool Equals(SLIST<T> other)
        {
            bool result = base.Equals(other);
            result &= other.Origin == null ? this.Origin == null : other.Origin.Equals(this.Origin);
            result &= other.Scale == null ? this.Scale == null : other.Scale.Equals(this.Scale);
            result &= this.m_digits == null ? other.m_digits == null : this.m_digits.Equals(other.m_digits);
            return result;
        }

        /// <summary>
        /// Determine equality
        /// </summary>
        public override bool Equals(object obj)
        {
            if (obj is SLIST<T>)
                return Equals((SLIST<T>)obj);
            return base.Equals(obj);
        }

        #endregion

        #region ISampledList Members

        /// <summary>
        /// Gets or sets the origin of the slist
        /// </summary>
        IQuantity ISampledList.Origin
        {
            get
            {
                return this.Origin;
            }
            set
            {
                this.Origin = (T)value;
            }
        }
     
        #endregion

      
    }
}
